tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/ILogger/DirectSubmission/Formatting/LoggerLogFormatter.cs

```diff
@@ -0,0 +1,160 @@
+                        // Theoretically we _should_ handle "duplicate" values, where the property has been added as state,
+                        // And then we try and add it again, but the backend seems to handle this ok
+                        if (scope is IEnumerable<KeyValuePair<string, object>> scopeItems)
+                        {
+                            foreach (var item in scopeItems)
+                            {
+                                var key = item.Key;
+                                haveSource |= LogFormatter.IsSourceProperty(key);
+                                haveService |= LogFormatter.IsServiceProperty(key);
+                                haveHost |= LogFormatter.IsHostProperty(key);
+                                haveTags |= LogFormatter.IsTagsProperty(key);
+                                haveEnv |= LogFormatter.IsEnvProperty(key);
+                                haveVersion |= LogFormatter.IsVersionProperty(key);
+
+                                LogFormatter.WritePropertyName(writer, key);
+                                LogFormatter.WriteValue(wrapper.Writer, item.Value);
+                            }
+                        }
+                        else
+                        {
+                            wrapper.Values.Add(scope); // add to list for inclusion in scope array
+                        }
+                    },
+                    writerWrapper);
+
+                if (writerWrapper.Values.Count > 0)
+                {
+                    writer.WritePropertyName("Scopes", escape: false);
+                    writer.WriteStartArray();
+
+                    for (var i = 0; i < writerWrapper.Values.Count; i++)
+                    {
+                        LogFormatter.WriteValue(writer, writerWrapper.Values[i]);
+                    }
+
+                    writer.WriteEndArray();
+                }
+            }
+
+            return new LogPropertyRenderingDetails(haveSource, haveService, haveHost, haveTags, haveEnv, haveVersion, messageTemplate);
+        }
+
+        private static string GetLogLevelString(int logLevel) =>
+            logLevel switch
+            {
+                0 => DirectSubmissionLogLevelExtensions.Verbose, // Trace
+                1 => DirectSubmissionLogLevelExtensions.Debug,
+                2 => DirectSubmissionLogLevelExtensions.Information,
+                3 => DirectSubmissionLogLevelExtensions.Warning,
+                4 => DirectSubmissionLogLevelExtensions.Error,
+                5 => DirectSubmissionLogLevelExtensions.Fatal, // Critical
+                _ => DirectSubmissionLogLevelExtensions.Unknown,
+            };
+
+        private readonly struct WriterWrapper
+        {
+            public readonly JsonWriter Writer;
+            public readonly List<object> Values;
+
+            public WriterWrapper(JsonWriter writer)
+            {
+                Writer = writer;
+                Values = new List<object>();
```

Comment:
It's really hard to know how often it will be populated, as it's likely app-dependent (I can't find anything obvious in aspnetcore repo that uses `BeginScope()` but doesn't set dictionary values for example). For some apps it will likely never have values, for others it will likely _always_ have values. 

Is it worth splitting the difference and making this a non-`readonly` struct, and lazy allocating the list? It's still a small struct, so the extra resulting copy should be pretty cheap, and would save a bit of allocation?

Reply:
Can't do that, because we don't control how `IExternalScopeProvider` is implemented. If they do something like:

```csharp
foreach (var scope in _scopes)
{
    callback(state);
}
```
Then each call to the callback will make a copy to the struct. It means we will end up with different lists in each callback.

Reply:
Ah, of course. I'd say lets go with the `class` approach, as guidance (from me anyway) would always be to use a dictionary, so `Values` wouldn't be used.

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/2240/commits/beb7dcd5f16b7f6ae225483d2bafea81b5a1d6fb